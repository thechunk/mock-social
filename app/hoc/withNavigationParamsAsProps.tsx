import React, {ComponentType, FunctionComponent} from 'react';
import {NavigationScreenProps, withNavigation} from "react-navigation";

/**
 * Applies navigation params from React Native Navigation to the wrapped component
 * @param Component - The component to wrap
 * @returns Wrapped component with props from navigation params
 */
type WrappedComponentProps = IWithNavigationParamsAsProps & NavigationScreenProps;
const withNavigationParamsAsProps = <P extends WrappedComponentProps>(Component: ComponentType<P>):
    FunctionComponent<WrappedComponentProps> =>
    (props: WrappedComponentProps) => {
        const Wrapped = withNavigation(Component);
        return <Wrapped {...props as P} {...props.navigation.state.params as P} />;
    };

export default withNavigationParamsAsProps;
