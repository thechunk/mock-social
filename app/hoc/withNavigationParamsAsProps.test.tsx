import * as React from 'react';
import withNavigationParamsAsProps from "./withNavigationParamsAsProps";
import {View} from "react-native";
import {shallow} from "enzyme";

describe('withNavigationParamsAsProps', () => {
    it('wrapper renders correctly', () => {
        const Component = () => (<View/>);
        const Wrapped: any = withNavigationParamsAsProps(Component);
        expect(shallow(<Wrapped navigation={{state: {params: {foo: 'bar'}}}} />)).toMatchSnapshot();
    });

    it('copies navigation params as props', () => {
        const Component = () => (<View />);
        const Wrapped: any = withNavigationParamsAsProps(Component);
        const wrappedComponent = shallow(<Wrapped navigation={{state: {params: {foo: 'bar'}}}}/>);
        expect(wrappedComponent.props().navigation.state.params.foo).toEqual('bar');
        expect(wrappedComponent.props().foo).toEqual('bar');
    });
});
