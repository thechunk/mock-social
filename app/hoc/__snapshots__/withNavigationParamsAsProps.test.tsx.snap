// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withNavigationParamsAsProps wrapper renders correctly 1`] = `
<withNavigation(Component)
  foo="bar"
  navigation={
    Object {
      "state": Object {
        "params": Object {
          "foo": "bar",
        },
      },
    }
  }
/>
`;
